addons/mail/static/tests/helpers/mock_server.js

```diff
@@ -2176,4 +2180,31 @@ MockServer.include({
             starred_counter: this._getRecords('mail.message', [['starred_partner_ids', 'in', user.partner_id]]).length,
         };
     },
+    /**
+     * Simulate the `notify_cancel_by_type` on `mail.thread` .
+     * Note that this method is overridden by snailmail module but not simulated here.
+     */
+    _mockMailThreadNotifyCancelByType(model, notificationType) {
+        // Query related mail messages and notifications
+        const notifMails = this._getRecords('mail.notification', [
+            ['notification_type', '=', notificationType],
+            ['notification_status', 'in', ['bounce', 'exception']],
+        ]).map(notification => ({
+            notif: notification,
+            mailMsg: this._getRecords('mail.message', [['id', '=', notification.mail_message_id]])[0]
+        })).filter(notifMail => notifMail.mailMsg.model === model && notifMail.mailMsg.author_id === this.currentPartnerId);
+        // Update notification status
+        this._mockWrite('mail.notification', [
+            notifMails.map(notifMail => notifMail.notif.id),
+            { notification_status: 'canceled' },
+        ]);
```

Comment:
I think the issue of this part of the code is trying to do 2 things at once.

It's mixing up the map that is supposed to generate the "result", from the filter to decide which notifications to consider.

I would have it like this instead:
1. get all the notifications as they are (no map), filter them (you can access the message itself from a _getRecords inside the filter directly, no need for a preliminary map), and put that in a `notifications` variable
2. write what you need on those notifications
3. iterate them to get their message id (you don't actually need the message record for this step, just `mail_message_id` key)